### DISCUSSION TITLE
How to run existing program with yaegi cli and extend program’s functionality

### DISCUSSION THREAD

**Discussion by KiranSatyaRaj** [traefik/yaegi#1505] [Q&A] [answered]
It is said that yaegi makes Go programs extensible, how can I test this on a program, in an interactive shell using yaegi cli, I made a simple program that gives the average which resides in the main(), and I want to extend the program by adding a new function that gives the median, now how can do this on an existing program using yaegi-cli, I'm just trying to learn what yaegi is, here the code of the program
```go
package main

import "fmt"

func main() {
	nums := []int{1, 2, 3, 4, 5}
	sum := 0
	for _, num := range nums {
		sum += num
	}
	average := float64(sum) / float64(len(nums))
	fmt.Println("The average is", average)
}
``` 
and in the terminal, I enter yaegi interactive shell, now I want to test/make changes with respect to this program, like adding a new function that calculates the median, here's the program
```go
func median(nums []int) float64 {
	sorted := make([]int, len(nums))
	copy(sorted, nums)
	sort.Ints(sorted)
	length := len(sorted)
	if length % 2 == 0 {
		return float64(sorted[length/2]+sorted[(length/2)-1]) / 2.0
	}
	return float64(sorted[length/2])
}
``` 
and run the `fmt.Println("The median is", median(nums))` without explicitly having to declare `nums`
this is how I did it
```console
$ yaegi run -i main.go
The average is 3
> func median(nums []int) float64 {
	sorted := make([]int, len(nums))
	copy(sorted, nums)
	sort.Ints(sorted)
	length := len(sorted)
	if length % 2 == 0 {
		return float64(sorted[length/2]+sorted[(length/2)-1]) / 2.0
	}
	return float64(sorted[length/2])
}
The average is 3
: 0xc000284330
> fmt.Println("The median is", median(nums))
main.go:1:64: undefined: nums
```

**Comment by mpl** [comment, marked as answer]
Hi.

what you're trying to do cannot work, because the nums variable from the main is local to the main function, it is not a global scope variable. So it is unknown when you're trying to give it as an argument to the median func(). It's as if you were trying to run that code:

```go
package main

import (
	"fmt"
	"sort"
)

func oldmain() {
	nums := []int{1, 2, 3, 4, 5}
	sum := 0
	for _, num := range nums {
		sum += num
	}
	average := float64(sum) / float64(len(nums))
	fmt.Println("The average is", average)
}

func median(nums []int) float64 {
	sorted := make([]int, len(nums))
	copy(sorted, nums)
	sort.Ints(sorted)
	length := len(sorted)
	if length%2 == 0 {
		return float64(sorted[length/2]+sorted[(length/2)-1]) / 2.0
	}
	return float64(sorted[length/2])
}

func main() {
	oldmain()
	fmt.Println("The median is", median(nums))
}
```

Here is how it would look if you fixed your main to have a globar var instead:

```console
% cat ./main.go 
package main

import (
	"fmt"
	"sort"
)

var nums = []int{1, 2, 3, 4, 5}

func main() {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	average := float64(sum) / float64(len(nums))
	fmt.Println("The average is", average)
}
% 
% 
% yaegi run -i ./main.go 
The average is 3
> func median(nums []int) float64 {
	sorted := make([]int, len(nums))
	copy(sorted, nums)
	sort.Ints(sorted)
	length := len(sorted)
	if length%2 == 0 {
		return float64(sorted[length/2]+sorted[(length/2)-1]) / 2.0
	}
	return float64(sorted[length/2])
}
The average is 3
: 0xc00020e2b0
> 
> fmt.Println("The median is", median(nums))
The median is 3
```

**Reply by KiranSatyaRaj** [reply]
it worked, thanks @mpl, I should've figured it out when I got this error message `main.go:1:64: undefined: nums`, I really appreciate for putting your time into this, now I'll explore more about yaegi.

**Reply by KiranSatyaRaj** [reply]
Hey @mpl, I have one quick question, what does this refer to `: 0xc00020e2b0` and also this simple program shows that we are modify it in real-time right?

**Reply by mpl** [reply]
off the top of my head I would say it's the address of the reflect Value for func median, or something like that.
But @mvertes knows better.

**Reply by mvertes** [reply]
@mpl you are correct :-)

**Reply by KiranSatyaRaj** [reply]
hey @mpl ,I'm going through the talk given at GoLab-2019 on yaegi by @mvertes, 'existing engines are less fit than Go itself`
does this mean that modern-day interpreters are usually built for dynamically typed languages, such as python, ruby, and javascript so they may not support the strictly typed system of Go, also being not able to meet Go's resource demands, like its support for concurrency and the current interpreters doesn't have good support it and also may not support to compilation process of Go, is the what it means?

**Reply by KiranSatyaRaj** [reply]
hey @mpl, by testing this new function median on the main.go, now when I'm running this in yaegi's REPL, can I say that the functions, variable values, or anything depending on the context, that are being tested in real-time in this REPL is/are local to the program that is provided in this command `yaegi run -i main.go`

**Reply by mpl** [reply]
Sorry, I do not understand what "local to the program" means, so I do not know.

**Reply by KiranSatyaRaj** [reply]
I was referring to the way I was able to extend the program without actually making changes to the existing program but doing it using yaegi-cli instead, making changes in real-time, and the code in the REPL knows that it's a part of the main.go file, I couldn't understand how the idea/implementation behind that works.

**Reply by KiranSatyaRaj** [reply]
hey @mpl , I cannot run the precompiled files, I get an error while doing so
```
yaegi run -i ./hello
./hello:1:1: illegal character U+007F
run: ./hello:1:1: illegal character U+007F
```

**Reply by mpl** [reply]
You're trying to run an interpreter on a compiled binary.

**Reply by KiranSatyaRaj** [reply]
> You're trying to run an interpreter on a compiled binary.

Yes, I'm sorry, what was I thinking

**Reply by KiranSatyaRaj** [reply]
Hey @mpl, I'm unable to run this command, I removed all the binaries and pkgs, for installing the new go version, now I couldn't install yaegi-cli, can you please help me through?
```
sudo curl -sfL https://raw.githubusercontent.com/traefik/yaegi/master/install.sh | bash -s -- -b $GOPATH/bin v0.9.0
traefik/yaegi info checking GitHub for tag 'v0.9.0'
traefik/yaegi info found version: 0.9.0 for v0.9.0/linux/amd64
install: cannot create regular file '/bin/yaegi': Permission denied
```
